> I am sometimes getting "You already have a parser..."

Yeah, it is running tests more than once, apparently when they begin with a 0.

```
TESTFILE=harness  ./runtests choose
../cpan/Archive-Tar/t/01_use.t .............. ok   
../cpan/Archive-Tar/t/02_methods.t .......... ok     
../cpan/Archive-Tar/t/03_file.t ............. ok     
../cpan/Archive-Tar/t/04_resolved_issues.t .. ok    
../cpan/Archive-Tar/t/05_iter.t ............. ok    
../cpan/Archive-Tar/t/06_error.t ............ ok    
../cpan/Archive-Tar/t/08_ptargrep.t ......... ok   
../cpan/Archive-Tar/t/09_roundtrip.t ........ ok   
../cpan/Archive-Tar/t/90_symlink.t .......... skipped: Skipping tests on this platform
../cpan/Archive-Tar/t/99_pod.t .............. skipped: Test::Pod v0.95 required for testing POD
../cpan/Archive-Tar/t/01_use.t .............. ok   
All tests successful.
Files=10, Tests=1240,  3 wallclock secs ( 0.35 usr  0.02 sys +  2.78 cusr  0.16 csys =  3.31 CPU)
Result: PASS
You already have a parser for (../cpan/Archive-Tar/t/01_use.t). Perhaps you have run the same test twice. at ../lib/TAP/Harness.pm line 597.
make: *** [makefile:865: test_harness] Error 25
```

also

```
TEST_JOBS=16 make test_harness -j16 TEST_ARGS='-re cpan/IO-Socket-IP'
```

it is somewhat random which one is triggered first what I can tell. For instance I did 

```
TEST_JOBS=16 make test_harness -j16 TEST_ARGS='-re cpan'
```

and I found the IO-Socket-IP case. Something wrong with the scheduler logic I guess. Ill try to find the bug, just posting this if you are still awake.

--- a/t/harness
+++ b/t/harness
@@ -178,6 +178,7 @@ if (@ARGV) {
         my %dir;
         my %serials;
         my %all_dirs;
+        my %map_file_to_dir;
 
         # Preprocess the list of tests
         for my $file (@last) {
@@ -199,6 +200,7 @@ if (@ARGV) {
 
                 $all_dirs{$path} = 1;
                 $serials{$path} = 1 if $name =~ / \A 0 /x;
+                $map_file_to_dir{$file} = $path;
             }
         }
 
@@ -217,34 +219,33 @@ if (@ARGV) {
               . "  Check your spelling:\n" . join "\n", @nonexistent_serials;
         }
 
-        # Remove the serial testing directories from the list of all
-        # directories.  The remaining ones are testable in parallel.  Make the
-        # parallel list a scalar with names separated by '|' so that below
-        # they will be added to a regular expression.
-        my $non_serials = join "|", grep { not exists $serials{$_} } keys %all_dirs;
+        for my $file (@last) {
+            my $file_dir = $map_file_to_dir{$file};
+
+            # Treat every file in each non-serial directory as its own
+            # "directory", so that it can be executed in parallel
+            if (! defined $serials{$file_dir}) {
+                $dir{$file} = { seq => $file };
+                $total_time{$file} = $times{$file} || 0;
+            }
+            else {  # Serial.  This file contributes time to the total needed
+                    # for its directory as a whole
+                $dir{$file_dir} = { seq => "$file_dir/*" };
+                $total_time{$file_dir} += $times{$file} || 0;
+            }
+        }
+
         undef %all_dirs;
         undef %serials;
 
-        for my $file (@last) {
-            # Treat every file in each non-serial directory as its own
-            # "directory", so that it can be executed in parallel
-            $file =~ m! \A ( (?: \.\. / )? (?: $non_serials )
-                         / [^/]+ \.t \z | .* [/] ) !x
-                or die "'$file'";
-            push @{$dir{$1}}, $file;
-
-            # This file contributes time to the total needed for the directory
-            # as a whole
-            $total_time{$1} += $times{$file} || 0;
-        }
         #print STDERR __LINE__, join "\n", sort sort_by_execution_order keys %dir
 
         push @tests, @last;
 
         # Generate T::H schedule rules that run the contents of each directory
         # sequentially.
-        push @seq, { par => [ map { s!/$!/*!; { seq => $_ } } sort
-                                                sort_by_execution_order keys %dir
+        push @seq, { par => [ map { $dir{$_} } sort sort_by_execution_order
+                                                                     keys %dir
                             ]
                    };
 

--- a/t/harness
+++ b/t/harness
@@ -45,18 +45,18 @@ sub _extract_tests {
     # would be as clear
     my @results;
     foreach (@_) {
-	my $ref = ref $_;
-	if ($ref) {
-	    if ($ref eq 'ARRAY') {
-		push @results, _extract_tests @$_;
-	    } elsif ($ref eq 'HASH') {
-		push @results, _extract_tests values %$_;
-	    } else {
-		die "Unknown reference type $ref";
-	    }
-	} else {
-	    push @results, glob $_;
-	}
+        my $ref = ref $_;
+        if ($ref) {
+            if ($ref eq 'ARRAY') {
+                push @results, _extract_tests @$_;
+            } elsif ($ref eq 'HASH') {
+                push @results, _extract_tests values %$_;
+            } else {
+                die "Unknown reference type $ref";
+            }
+        } else {
+            push @results, glob $_;
+        }
     }
     @results;
 }
@@ -121,10 +121,10 @@ if (@ARGV) {
     # but for now, run all directories in sequence.
 
     unless (@tests) {
-	my @seq = <base/*.t>;
+        my @seq = <base/*.t>;
 
         my @last;
-	my @next = qw(comp run cmd);
+        my @next = qw(comp run cmd);
 
         # The remaining core tests are either intermixed with the non-core for
         # more parallelism (if PERL_TEST_HARNESS_ASAP is set non-zero) or done
@@ -132,58 +132,58 @@ if (@ARGV) {
         my $which = $ENV{PERL_TEST_HARNESS_ASAP} ? \@last : \@next;
 
         push @$which, qw(io re opbasic op uni mro lib porting perf);
-	push @$which, 'japh' if $torture;
-	push @$which, 'win32' if $^O eq 'MSWin32';
-	push @$which, 'benchmark' if $ENV{PERL_BENCHMARK};
-	push @$which, 'bigmem' if $ENV{PERL_TEST_MEMORY};
+        push @$which, 'japh' if $torture;
+        push @$which, 'win32' if $^O eq 'MSWin32';
+        push @$which, 'benchmark' if $ENV{PERL_BENCHMARK};
+        push @$which, 'bigmem' if $ENV{PERL_TEST_MEMORY};
 
-	# Hopefully TAP::Parser::Scheduler will support this syntax soon.
-	# my $next = { par => '{' . join (',', @next) . '}/*.t' };
-	my $next = { par => [
-			     map { "$_/*.t" } @next
-			    ] };
-	@tests = _extract_tests ($next);
+        # Hopefully TAP::Parser::Scheduler will support this syntax soon.
+        # my $next = { par => '{' . join (',', @next) . '}/*.t' };
+        my $next = { par => [
+                             map { "$_/*.t" } @next
+                            ] };
+        @tests = _extract_tests ($next);
 
-	my $last = { par => '{' . join (',', @last) . '}/*.t' };
-	@last = _extract_tests ($last);
+        my $last = { par => '{' . join (',', @last) . '}/*.t' };
+        @last = _extract_tests ($last);
 
-	# This is a bit of a game, because we only want to sort these tests in
-	# speed order. base/*.t wants to run first, and ext,lib etc last and in
-	# MANIFEST order
-	if ($jobs) {
-	    require App::Prove::State;
-	    $state = App::Prove::State->new({ store => 'test_state' });
-	    $state->apply_switch('slow', 'save');
-	    # For some reason get_tests returns *all* the tests previously run,
-	    # (in the right order), not simply the selection in @tests
-	    # (in the right order). Not sure if this is a bug or a feature.
-	    # Whatever, *we* are only interested in the ones that are in @tests
-	    my %seen;
-	    @seen{@tests} = ();
-	    @tests = grep {exists $seen{$_} } $state->get_tests(0, @tests);
-	}
-	@tests = (@seq, @tests);
-	push @seq, $next;
+        # This is a bit of a game, because we only want to sort these tests in
+        # speed order. base/*.t wants to run first, and ext,lib etc last and in
+        # MANIFEST order
+        if ($jobs) {
+            require App::Prove::State;
+            $state = App::Prove::State->new({ store => 'test_state' });
+            $state->apply_switch('slow', 'save');
+            # For some reason get_tests returns *all* the tests previously run,
+            # (in the right order), not simply the selection in @tests
+            # (in the right order). Not sure if this is a bug or a feature.
+            # Whatever, *we* are only interested in the ones that are in @tests
+            my %seen;
+            @seen{@tests} = ();
+            @tests = grep {exists $seen{$_} } $state->get_tests(0, @tests);
+        }
+        @tests = (@seq, @tests);
+        push @seq, $next;
 
-	push @last,
-	    _tests_from_manifest($Config{extensions}, $Config{known_extensions});
-	my %times;
-	if ($state) {
-	    # Where known, collate the elapsed times by test name
-	    foreach ($state->results->tests()) {
-		$times{$_->name} = $_->elapsed();
-	    }
-	}
+        push @last,
+          _tests_from_manifest($Config{extensions}, $Config{known_extensions});
+        my %times;
+        if ($state) {
+            # Where known, collate the elapsed times by test name
+            foreach ($state->results->tests()) {
+                $times{$_->name} = $_->elapsed();
+            }
+        }
 
-	my %dir;
+        my %dir;
         my %serials;
         my %all_dirs;
 
         # Preprocess the list of tests
-	for my $file (@last) {
-	    if ($^O eq 'MSWin32') {
-		$file =~ s,\\,/,g; # canonicalize path
-	    };
+        for my $file (@last) {
+            if ($^O eq 'MSWin32') {
+                $file =~ s,\\,/,g; # canonicalize path
+            };
 
             # Keep a list of the distinct directory names, and another list of
             # those which contain a file whose name begins with a 0
@@ -225,30 +225,30 @@ if (@ARGV) {
         undef %all_dirs;
         undef %serials;
 
-	for my $file (@last) {
+        for my $file (@last) {
             # Treat every file in each non-serial directory as its own
             # "directory", so that it can be executed in parallel
             $file =~ m! \A ( (?: \.\. / )? (?: $non_serials )
                          / [^/]+ \.t \z | .* [/] ) !x
                 or die "'$file'";
-	    push @{$dir{$1}}, $file;
+            push @{$dir{$1}}, $file;
 
             # This file contributes time to the total needed for the directory
             # as a whole
-	    $total_time{$1} += $times{$file} || 0;
-	}
+            $total_time{$1} += $times{$file} || 0;
+        }
         #print STDERR __LINE__, join "\n", sort sort_by_execution_order keys %dir
 
-	push @tests, @last;
+        push @tests, @last;
 
-	# Generate T::H schedule rules that run the contents of each directory
-	# sequentially.
-	push @seq, { par => [ map { s!/$!/*!; { seq => $_ } } sort
+        # Generate T::H schedule rules that run the contents of each directory
+        # sequentially.
+        push @seq, { par => [ map { s!/$!/*!; { seq => $_ } } sort
                                                 sort_by_execution_order keys %dir
                             ]
                    };
 
-	$rules = { seq => \@seq };
+        $rules = { seq => \@seq };
     }
 }
 if ($^O eq 'MSWin32') {
@@ -300,71 +300,71 @@ my $h = TAP::Harness->new({
     verbosity   => $Verbose,
     timer       => $ENV{HARNESS_TIMER},
     exec        => sub {
-	my ($harness, $test) = @_;
+        my ($harness, $test) = @_;
 
-	my $options = $options{$test};
-	if (!defined $options) {
-	    $options = $options{$test} = _scan_test($test, $type);
-	}
+        my $options = $options{$test};
+        if (!defined $options) {
+            $options = $options{$test} = _scan_test($test, $type);
+        }
 
-	(local $Valgrind_Log = "$test.valgrind-current") =~ s/^.*\///;
+        (local $Valgrind_Log = "$test.valgrind-current") =~ s/^.*\///;
 
-	return [ split ' ', _cmd($options, $type) ];
+        return [ split ' ', _cmd($options, $type) ];
     },
 });
 
 # Print valgrind output after test completes
 if ($ENV{PERL_VALGRIND}) {
     $h->callback(
-		 after_test => sub {
-		     my ($job) = @_;
-		     my $test = $job->[0];
-		     my $vfile = "$test.valgrind-current";
-	             $vfile =~ s/^.*\///;
+                 after_test => sub {
+                     my ($job) = @_;
+                     my $test = $job->[0];
+                     my $vfile = "$test.valgrind-current";
+                     $vfile =~ s/^.*\///;
 
-		     if ( (! -z $vfile) && open(my $voutput, '<', $vfile)) {
-			print "$test: Valgrind output:\n";
-			print "$test: $_" for <$voutput>;
-			close($voutput);
-		     }
+                     if ( (! -z $vfile) && open(my $voutput, '<', $vfile)) {
+                        print "$test: Valgrind output:\n";
+                        print "$test: $_" for <$voutput>;
+                        close($voutput);
+                     }
 
-		     (local $Valgrind_Log = "$test.valgrind-current") =~ s/^.*\///;
+                     (local $Valgrind_Log = "$test.valgrind-current") =~ s/^.*\///;
 
-		     _check_valgrind(\$htoolnm, \$hgrind_ct, \$test);
-		 }
-		 );
+                     _check_valgrind(\$htoolnm, \$hgrind_ct, \$test);
+                 }
+                 );
 }
 
 if ($state) {
     $h->callback(
-		 after_test => sub {
-		     $state->observe_test(@_);
-		 }
-		 );
+                 after_test => sub {
+                     $state->observe_test(@_);
+                 }
+                 );
     $h->callback(
-		 after_runtests => sub {
-		     $state->commit(@_);
-		 }
-		 );
+                 after_runtests => sub {
+                     $state->commit(@_);
+                 }
+                 );
 }
 
 $h->callback(
-	     parser_args => sub {
-		 my ($args, $job) = @_;
-		 my $test = $job->[0];
-		 _before_fork($options{$test});
-		 push @{ $args->{switches} }, "-I../../lib";
-	     }
-	     );
+             parser_args => sub {
+                 my ($args, $job) = @_;
+                 my $test = $job->[0];
+                 _before_fork($options{$test});
+                 push @{ $args->{switches} }, "-I../../lib";
+             }
+             );
 
 $h->callback(
-	     made_parser => sub {
-		 my ($parser, $job) = @_;
-		 my $test = $job->[0];
-		 my $options = delete $options{$test};
-		 _after_fork($options);
-	     }
-	     );
+             made_parser => sub {
+                 my ($parser, $job) = @_;
+                 my $test = $job->[0];
+                 my $options = delete $options{$test};
+                 _after_fork($options);
+             }
+             );
 
 my $agg = $h->runtests(@tests);
 _cleanup_valgrind(\$htoolnm, \$hgrind_ct);

--- a/t/harness
+++ b/t/harness
@@ -39,6 +39,36 @@ my (@tests, @re, @anti_re);
 # [.VMS]TEST.COM calls harness with empty arguments, so clean-up @ARGV
 @ARGV = grep $_ && length( $_ ) => @ARGV;
 
+while ($ARGV[0] && $ARGV[0]=~/^-(n?)re/) {
+    my $ary= $1 ? \@anti_re : \@re;
+
+    if ( $ARGV[0] !~ /=/ ) {
+        shift @ARGV;
+        while (@ARGV and $ARGV[0] !~ /^-/) {
+            push @$ary, shift @ARGV;
+        }
+    } else {
+        push @$ary, (split/=/,shift @ARGV)[1];
+    }
+}
+
+my $jobs = $ENV{TEST_JOBS};
+my ($rules, $state, $color);
+
+if ($ENV{HARNESS_OPTIONS}) {
+    for my $opt ( split /:/, $ENV{HARNESS_OPTIONS} ) {
+        if ( $opt =~ /^j(\d*)$/ ) {
+            $jobs ||= $1 || 9;
+        }
+        elsif ( $opt eq 'c' ) {
+            $color = 1;
+        }
+        else {
+            die "Unknown HARNESS_OPTIONS item: $opt\n";
+        }
+    }
+}
+
 sub _extract_tests;
 sub _extract_tests {
     # This can probably be done more tersely with a map, but I doubt that it
@@ -59,35 +89,6 @@ sub _extract_tests {
         }
     }
     @results;
-}
-
-while ($ARGV[0] && $ARGV[0]=~/^-(n?)re/) {
-    my $ary= $1 ? \@anti_re : \@re;
-
-    if ( $ARGV[0] !~ /=/ ) {
-        shift @ARGV;
-        while (@ARGV and $ARGV[0] !~ /^-/) {
-            push @$ary, shift @ARGV;
-        }
-    } else {
-        push @$ary, (split/=/,shift @ARGV)[1];
-    }
-}
-
-my $jobs = $ENV{TEST_JOBS};
-my ($rules, $state, $color);
-if ($ENV{HARNESS_OPTIONS}) {
-    for my $opt ( split /:/, $ENV{HARNESS_OPTIONS} ) {
-        if ( $opt =~ /^j(\d*)$/ ) {
-            $jobs ||= $1 || 9;
-        }
-        elsif ( $opt eq 'c' ) {
-            $color = 1;
-        }
-        else {
-            die "Unknown HARNESS_OPTIONS item: $opt\n";
-        }
-    }
 }
 
 my %total_time;

